Add field type for version strings

cbuescher opened the pull request:
This PR adds a new 'version' field type that allows indexing string values
representing software versions similar to the ones defined in the Semantic
Versioning definition (semver.org). The field behaves very similar to a
'keyword' field but allows efficient sorting and range queries that take into
accound the special ordering needed for version strings. For example, the main
version parts are sorted numerically (ie 2.0.0 < 11.0.0) whereas this wouldn't
be possible with 'keyword' fields today.

Valid version values are similar to the Semantic Versioning definition, with the
notable exception that in addition to the "main" version consiting of
major.minor.patch, we allow less or more than three numeric identifiers, i.e.
"1.2" or "1.4.6.123.12" are treated as valid too.

Relates to #48878

Commit:
Add field type for version strings

This PR adds a new 'version' field type that allows indexing string values
representing software versions similar to the ones defined in the Semantic
Versioning definition (semver.org). The field behaves very similar to a
'keyword' field but allows efficient sorting and range queries that take into
accound the special ordering needed for version strings. For example, the main
version parts are sorted numerically (ie 2.0.0 < 11.0.0) whereas this wouldn't
be possible with 'keyword' fields today.

Valid version values are similar to the Semantic Versioning definition, with the
notable exception that in addition to the "main" version consiting of
major.minor.patch, we allow less or more than three numeric identifiers, i.e.
"1.2" or "1.4.6.123.12" are treated as valid too.

Relates to #48878

elasticmachine commented:
Pinging @elastic/es-search (:Search/Mapping)

Commit:
Add yaml for search, ranges and scripting

Commit:
Add brute force regex query support

Commit:
Add fuzzy query support

Commit:
Merge branch 'master' into add-version-field

Commit:
Remove some left-over commented code

Commit:
Adding tests

mayya-sharipova commented on x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java:
```diff
@@ -0,0 +1,550 @@
+                } else if (propName.equals("store_malformed")) {
+                    builder.storeMalformed(XContentMapValues.nodeBooleanValue(propNode, name + ".store_malformed"));
+                    iterator.remove();
+                } else if (TypeParsers.parseMultiField(builder::addMultiField, name, parserContext, propName, propNode)) {
```
What's the need of  extra `TypeParsers.parseMultiField` on line 169 if `TypeParsers.parseField` on line 155 already takes care of multi-field?   Should we choose one of these 2 options?

mayya-sharipova commented on x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java:
```diff
@@ -0,0 +1,550 @@
+
+        protected String nullValue = Defaults.NULL_VALUE;
+        private boolean storeMalformed = false;
+
```
Should we override and disallow `indexOptions` in Builder, similarly how NumberFieldMapper#Builder does it?

mayya-sharipova commented on x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java:
```diff
@@ -0,0 +1,550 @@
+            if (parser.currentToken() == XContentParser.Token.VALUE_NULL) {
+                versionString = nullValue;
+            } else {
+                versionString = parser.textOrNull();
```
Do we want to assign `nullValue` here as as well in cases where a string is empty or null?

cbuescher replied:
I agree that default values are probably not that meaningful here, but e.g. `0.0.0` might be something users might want to index in case this value is null. I think I added this for versions so they behave as close to "keyword" fields as possible, also on the indexing side, but I'm open to discuss and remove this if you think its not required.

jtibshirani replied:
It would be great to discuss (maybe through a different channel?), I've been wondering in general about whether most field types should support `null_value`.

cbuescher replied:
I changed this PR to not allow setting `null_value`, however we accept `null` as a field value in the input json in which case we don't index any value into the field.

jtibshirani reviewed (commented):
I was getting up to speed with the feature so I could help with reviews, and left some small comments. I haven't yet reviewed in detail though.

Commit:
Make string_stats work

Commit:
Merge branch 'master' into add-version-field

mayya-sharipova commented on x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionEncoder.java:
```diff
@@ -0,0 +1,220 @@
+                }
+                boolean isNumeric = preReleasePart.chars().allMatch(x -> Character.isDigit(x));
+                if (isNumeric) {
+                    prefixDigitGroupsWithLength(preReleasePart, encodedBytes);
```
may be worth separating this function into 2 functions: one that just prefix digit group with length and another one extracts main versions?

cbuescher replied:
I couldn't find a simple way to split this methodf that wouldn't involve running through the input string at least twice. This way we avoid two passes, but I'm happy for any suggestions how to change this if you have an idea...

cbuescher commented on x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionEncoder.java:
```diff
@@ -0,0 +1,220 @@
+
+        // don't treat non-legal versions further, just mark them as illegal and return
+        if (legalVersionString(versionParts) == false) {
+            return new EncodedVersion(new BytesRef(versionString), false, true, 0, 0, 0);
```
I made the class package private and the ctor private, hope that fits your suggestion.

jtibshirani replied:
👍  this helps address the concern.

mayya-sharipova commented on x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionEncoder.java:
```diff
@@ -0,0 +1,220 @@
+        );
+    }
+
+    private static Integer[] prefixDigitGroupsWithLength(String input, BytesRefBuilder result) {
```
That's quite a clever encoding! Nice work! What I like about it is that it allows the decoding to the original values.

Another way to do encoding I was thinking of is to standardize main versions to 3 bytes (1 byte of each component), but standardization would not allow the decoding to the original values.

cbuescher replied:
Yes, being able to get back the original values is quite important here, e.g. we need that when needing to display encoded docvalues (e.g. keys for terms aggs where I first encountered that problem). I tried several approaches with encoding main verison parts as bytes (several widths) but they all failed either because bein too restrictive (we also want to allow less or more than 3 main digits, as an extension to strict semver, e.g. "1.0.14.20201027") and the range of verison numbers is in theory unlimited. Thats why in the end I prefer this simpler apporoach which can also be used for numeric parts in the pre-release part of the version string.

cbuescher commented on x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionEncoder.java:
```diff
@@ -0,0 +1,220 @@
+ *  <li> a numeric major.minor.patch part starting the version string (e.g. 1.2.3)
+ *  <li> an optional "pre-release" part that starts with a `-` character and can consist of several alpha-numerical sections
+ *  separated by dots (e.g. "-alpha.2.3")
+ *  <li> an optional "build" part that starts with a `+` character. This will simply be treated as a prefix with no guaranteed ordering,
```
Sorry for the confusion, it should be "suffix" and I changed the description to a hopefully less confusing "ascii order"

Commit:
Adressing review comments

cbuescher commented:
@mayya-sharipova @jtibshirani thanks for the reviews, I tried adressing parts of them with another commit. The question around weather `null_value` should be supported by this fieldtype is an interesting one, I left a comment why I used it coming from the assumtion we want to behave as closely as "keyword" as possible, but that might have downsides here. Let me know what you think.

Commit:
Remove 'store_malformed' option

Commit:
Add tests for handling empty string

cbuescher commented:
Since @jtibshirani and some others have mentioned that they think the 'store_malformed' option might not be needed but that we should always accept version strings even if invalid under strict semver, I removed that option with bdb89a2. This means invalid versions like "1.e42.123" are always accepted without error and can be retrieved with exact searches, they will also turn up in aggregations etc... but they will be sorted _after_ all valid versions. For me this also brought up the question where we would sort the empty String `""`, since its invalid, and I added a couple of tests that make sure we sort an input like that after the valid versions. This can come in handy, since users can filter all "invalid" versions by using the empty string as a lower bound in a range query this way.

Commit:
fix yaml test

Commit:
Merge branch 'master' into add-version-field

mayya-sharipova commented on x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java:
```diff
@@ -0,0 +1,526 @@
+            context.doc().add(field);
+            // encode the first 16 bytes as points for efficient range query
+            byte[] first16bytes = Arrays.copyOfRange(encodedVersion.bytes, encodedVersion.offset, 16);
+            context.doc().add(new BinaryPoint(fieldType().name(), first16bytes));
```
Should we index `BinaryPoint` only when `indexOptions() != IndexOptions.NONE `(because currently this field will be indexed even when `IndexOptions.NONE` and stored=true) ?

mayya-sharipova commented on x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java:
```diff
@@ -0,0 +1,526 @@
+            byte[] first16bytes = Arrays.copyOfRange(encodedVersion.bytes, encodedVersion.offset, 16);
+            context.doc().add(new BinaryPoint(fieldType().name(), first16bytes));
+        }
+        context.doc().add(new SortedSetDocValuesField(fieldType().name(), encodedVersion));
```
Should we create this field only when hasDocValues() == true?

cbuescher replied:
Sure. I was also thinking about always setting "docValue == true" for this field regardless of the users settings. I think @jpountz suggested this at some point during reviewing the draft, maybe he can restate this opinion here?

cbuescher replied:
I updated the PR so we now don't allow disabling `doc_values` for this specialized field, so it will always be `true`.

mayya-sharipova commented on x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java:
```diff
@@ -0,0 +1,526 @@
+        }
+        context.doc().add(new SortedSetDocValuesField(fieldType().name(), encodedVersion));
+
+        // add additional information extracted from version string
```
Should we create all the fields below only when `indexOptions() != IndexOptions.NONE ` and `SortedNumericDocValuesField` when hasDocValues() == true?

cbuescher replied:
I updated the PR so we now don't allow disabling doc_values or setting specific index options for this specialized field.

mayya-sharipova commented on x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java:
```diff
@@ -0,0 +1,526 @@
+
+        @Override
+        public Query rangeQuery(Object lowerTerm, Object upperTerm, boolean includeLower, boolean includeUpper, QueryShardContext context) {
+            if (context.allowExpensiveQueries() == false) {
```
I am wondering if this query still should be considering expensive if with  a binaryPoint query we can efficiently narrow down terms?

mayya-sharipova replied:
I was also wondering since lowerTerm and upperTerm represent valid versions can we extract major, minor, patch fields  and use them for querying to further narrow down terms?

cbuescher replied:
lower and upper don't necessarily represent valid versions, we'd also be able to use non-semver versions that then sort above the valid ones. Major/Minor/Patch might be not available for them.  As for the expensive part, I'm happy to change, I'm not familiar enough with the cost of binaryPoint so I'd follow your judgement or if anyone else has an opinion on this.

mayya-sharipova reviewed (commented):
@cbuescher Thanks for the PR, exciting feature!  I have done the initial review. I am interested to see documentation or more tests that demonstrate the usage of other fields `major`, `minor`,  `patch` and `prerelease`.

cbuescher commented:
@mayya-sharipova thanks for your review, I will address your comments and add the tests you asked for. Would you prefer adding docs in this PR or in a follow up?

jimczi commented on x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java:
```diff
@@ -0,0 +1,526 @@
+        EncodedVersion encoding = encodeVersion(versionString);
+        BytesRef encodedVersion = encoding.bytesRef;
+        if (fieldType.indexOptions() != IndexOptions.NONE || fieldType.stored()) {
+            Field field = new Field(fieldType().name(), encodedVersion, fieldType);
```
That seems redundant. If I understand correctly, the encoded version is stored in three structures:

- Fully encoded version in `Postings`

- The first 16 bytes of the encoded version in `Points`

- Fully encoded version in `SortedSetDocValues`

The main benefit of indexing `Points` is that you don't need the postings. We can use them as an approximation and rely on the doc values to validate matches in a range query. Prefix, wildcard, regex are more tricky to handle with `Points` so I guess that is why you also create an indexed `Field`. However, I think we should only pick one strategy. The current approach is too costly imo, we should control the indexing cost more carefully.

jimczi commented on x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java:
```diff
@@ -0,0 +1,526 @@
+
+            return new BooleanQuery.Builder().add(new BooleanClause(pointPrefixQuery, Occur.MUST))
+                .add(new BooleanClause(termQuery, Occur.MUST))
+                .build();
```
The two queries that you mixed are totally independent so the resulting boolean query is just more costly than a plain `TermRangeQuery`. 
The idea behind using a prefix query on `Points` is that you can use the doc values to validate the matches. You can check how the `wildcard` field [does](https://github.com/elastic/elasticsearch/blob/2c501a580c857c232d96195b410040f063b0169a/x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java#L726)  since it's very similar. The `AutomatonQueryOnBinaryDv` leverages the `TwoPhaseIterator` to only validate documents found by other filter (the point range query approximation). 
As I said in previous comments, I think we need to make a choice. Relying on `Points` or `Postings` is ok but not both.

cbuescher replied:
Thanks, I looked into this and hope I understood things right. I pushed cfe2dba which follows a simiar approach as AutomatonQueryOnBinaryDv but works on SortedSet doc values.

cbuescher replied:
After talking again about this I changed the range validation query again in eaa1ef8. Since we also talked about not using Points at all I will use this version when comparing to a solution that doesn't use Points at all (a simple TermRangeQuery)

jimczi reviewed (commented):
I did a first round of review and left some questions and comments. The current strategy to index the field seems redundant. If we want to handle all types of query (`fuzzy`, `prefix`,  `wildcard`, `regex`) I'd lean towards using only Postings. However we can also consider the field as an ordering scheme that only accepts exact and `range` queries and in such case indexing only Points makes sense.
I don't have a clear preference so I am curious to hear what others think ?

mayya-sharipova commented:
> The current strategy to index the field seems redundant.

I was also thinking the same, that's why I asked @cbuescher  for additional documentation and query examples to clarify what these  additional fields allow us to achieve.

cbuescher commented on x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java:
```diff
@@ -0,0 +1,526 @@
+
+    private void addVersionPartSubfield(ParseContext context, String fieldName, Integer versionPart) {
+        if (versionPart != null) {
+            context.doc().addAll(NumberType.INTEGER.createFields(fieldName, versionPart, true, true, false));
```
This is not aimed at making the range queries faster like in indexing the 16 byte prefix as point. I added these fields because the telemetry team have the use case that they want to filter and aggregate on mayor / minor version (maybe patch), e.g. with histograms or terms aggregations. Think along the lines of "give me a breakdown of all mayor/minor/patch" versions running in my cluster. That information would otherwise have to be parsed out again on aggregations e.g. by scripts, so adding these field will make these kind of aggregations faster. I was considering having to switch these additional subfields on with a type parameter so its not done by default.

jimczi replied:
Ok thanks for explaining. Although I think we should discuss more alternatives to handle these requirements. Extracting the major/minor/patch shouldn't be considered costly in aggregations for instance. We also have plenty of ways to make the decoding faster if needed.
Querying all `0` minor should also  be possible with the full version so I'd prefer to find an alternative that doesn't require 3 additional fields.

cbuescher commented on x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java:
```diff
@@ -0,0 +1,526 @@
+        @Override
+        public Mapper.Builder<?> parse(String name, Map<String, Object> node, ParserContext parserContext) throws MapperParsingException {
+            Builder builder = new Builder(name);
+            TypeParsers.parseField(builder, name, node, parserContext);
```
Would that include common options like "meta" or "boost"? It would certainly simplify things if we would had fixed 'index_options' or 'doc_values'  settings. My impression was that people would like to use this like a keyword field and expect certain basic options.

cbuescher replied:
I updated the PR so the mapper now extends ParametrizedFieldMapper, which limits the number of options. I only added parsing for "meta" field for now since we seem to want to support this for all fields.

cbuescher commented on x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java:
```diff
@@ -0,0 +1,526 @@
+
+        EncodedVersion encoding = encodeVersion(versionString);
+        BytesRef encodedVersion = encoding.bytesRef;
+        if (fieldType.indexOptions() != IndexOptions.NONE || fieldType.stored()) {
```
Happy to not allow configurable ' index_options' and e.g. also assume 'doc_values'  are always turned on (if we decide thats where to store the full encoded version)

cbuescher replied:
`index_options` are now fixed with a recent update

mayya-sharipova commented on x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java:
```diff
@@ -0,0 +1,526 @@
+        return value.toString();
+    }
+
+    private static DocValueFormat VERSION_DOCVALUE = new DocValueFormat() {
```
Do you need to register this format in `NamedWriteableRegistry` like we register other format in `SearchModule#registerValueFormats`?

cbuescher replied:
Good point, I guess this hasn't shown up in tests yet because they are mostly single node. I'll see if this would e.g. fail a yaml aggs test. I can probably register this in the plugin.

cbuescher replied:
I added c70e7c1 to register the format and an IT test that would have caught this when serializing aggregations across nodes. Thanks for catching this!

Commit:
Change range query to points approximation and dv validation

Commit:
Register DocValueFormat via plugin

Commit:
Merge branch 'master' into add-version-field

Commit:
Merge branch 'master' into add-version-field

Commit:
Add valueFetcher() method

cbuescher commented:
Thanks everybody for the reviews so far. I see three main open discussion topics which I’ll try to summarise and collect a few options for:

* Should we control  `index_options`  and `doc_values` or allow the user to configure them? The majority of comments seem to indicate we would like to always enable `doch_values` and user fixed `index_options` since this is a specialised type. We should also limit the number of allowed options for the field type in general. This probably means we also shouldn’t allow `null_values`. Looking at the new `ParametrizedFieldMapper` it looks like it might make sense extending this new field type from that as well.

* The current approach of storing fully encoded version in Postings, a 16 byte prefix as Points and an encoded version in SortedSetDocValues seems too costly. The suggestions is to use Points as an approximation and rely on the doc values to validate matches in a range query.
	* Caveats: I’m not yet sure how to make Prefix, wildcard and regex query work efficiently with only points and doc values. This needs some further examination. Dropping doc_values instead seems not possible to me since they are required for sorting etc. which this field type is mainly intended for

* The cost of storing the `preRelease`, `mayor`, `minor` and `patch` subfields that could be used for fast filtering for e.g. “all released versions” / all pre-release versions or to aggregate on mayor or mayor/minor versions (e.g. in histograms) is deemed to high. Again I’m not sure how to efficiently solve this without these subfields yet but we should discuss the use-cases with the requesting users again and see if they can be solved in some other way without storing the subfields.

Commit:
Merge branch 'master' of github.com:elastic/elasticsearch into add-version-field

Commit:
Restrict field options by moving to ParametrizedFieldMapper

Commit:
Merge branch 'master' into add-version-field

Commit:
Change range query validation part

cbuescher commented:
I moved the new fields mapper to extend ParametrizedFieldMapper and removed support for most options except the 'meta'  field for now. This should address a few of the earlier comments about support for `null_value` and also fixes the index option to always index and add doc_values which we need for sorting.

cbuescher commented:
Regarding the choice of Lucene structures we index to brought up in https://github.com/elastic/elasticsearch/pull/59773#discussion_r462949258 I did some experiments using the disk usage tool to compare space requirements using Postings and Points. 
Since we expect values stored in the `version` field to have a relatively low cardinality, I used the ~300 version tags of Elasticsearch as a base and augmented these versions with random suffixes to derive two more datasets with ~3000 and ~30000 unique versions. I indexed 1 Mio documents drawing randomly from the set of versions.

In a first experiment I assumed a uniform distribution of documents across available versions. Here's the disk usage tools measurement for the version field (in bytes):

| unique terms 	| terms dict   	| postings  	| terms+postings 	| points    	| docvalues 	|
| ---	| ---	| --- | ---	| ---	| ---	|
| 300          	| 4.115        	| 1.462.482 	| 1.466.597      	| 1.894.728 	| 1.501.859 	|
| 3000         	| 37.568       	| 1.888.442 	| 1.926.010      	| 2.979.954 	| 1.515.882 	|
| 30000        	| 218.479      	| 2.565.601 	| 2.784.080      	| 3.158.028 	| 2.130.482 	|

In a second experiment the distribution of versions if heavily skewed towards higher versions, trying to emulate e.g. monitoring systems where most of the versions running in production are of a relatively recent version:

| unique terms 	| terms dict   	| postings  	| terms+postings 	| points    	| docvalues 	|
| ---	| ---	| --- | ---	| ---	| ---	|
| 300          	| 4.163       	| 1.304.365 	| 1.308.528      	| 1.649.510 	| 1.501.862 	|
| 3000         	| 35.168       	| 1.727.602 	| 1.762.770      	| 2.442.291 	| 1.516.013 	|
| 30000        	| 201.809      	| 2.288.765 	| 2.490.574      	| 3.118.415 	| 2.130.021 	|

In both scenarios and for varying dictionary sizes it looks the storage cost for the dictionary and postings (terms+postings) is always lower than for points.

I'll continue this investigation looking at query performance for range queries using either just postings or points.

cbuescher commented:
Some preliminary query performance experiments, comparing average time for 1000 random range queries on the version field using (a) TermRangeQuery on Postings only vs. (b) combined query on 16 byte points prefix and slow validation query on doc values. Test run locally on the three datasets from above (1Mio docs, dictionary sizes 300, 3000, 30000). 20 repetitions of 1000 random range queries gives:

Postings (TermRangeQuery):

| unique terms 	| avg. query time ms |
| ---	| ---	| 
| 300          	| 2580        	| 
| 3000         	| 3150       	|
| 30000        	| 5058     	|

Points + Doc Values:

| unique terms 	| avg. query time ms |
| ---	| ---	| 
| 300          	| 2658        	| 
| 3000         	| 2923       	|
| 30000        	| 5069     	|

From these results it looks like the two implementations very similar in terms of speed for these dictionary sizes. I will try to confirm this by running a few more experiments using Rally.

jimczi commented on x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java:
```diff
@@ -0,0 +1,483 @@
+            // point query on the 16byte prefix
+            Query pointPrefixQuery = BinaryPoint.newRangeQuery(name(), lowerBytes, upperBytes);
+
+            Query validationQuery = SortedSetDocValuesField.newSlowRangeQuery(name(), lower, upper, includeLower, includeUpper);
```
You don't need to add the validation query if the lower and upper values have less than 17 bytes ?

jimczi replied:
The validation can also be avoided on values that have less than 17 bytes so you'll maybe need a custom validation query ?

cbuescher replied:
Good point, I probably still want to use it if includeLower or includeUpper is `false` since the points query will always contain the edges as far as I've seen from previous tests. I will push an update on this before I continue with testing vs. TermRangeQuery using only postings.

Commit:
Small improvement to range query using points

Commit:
Merge branch 'master' into add-version-field

cbuescher commented:
Continuing with performance experiments, I ran rally on a range query implementation using only Postings (TermRangeQuery) as "baseline" and comparing to the current implementaion using Point/Docvalues (contender). While Points might to be slightly faster in some occasions, the amount seems pretty small and in the 99th percentile range there are even slower cases (although this might be outliers. 
Setup for all experiments: using differing dictionaly sizes (300, 3000, 30000) on 1 Mio. docs, running range queries with random versions on it. Leaving out the 99.9 and 100 percentile for better readability:

|                                                        Metric |                   Task |    Baseline |   Contender |     Diff |   Unit |
|--------------------------------------------------------------:|-----------------------:|------------:|------------:|---------:|-------:|
|                                  50th percentile service time |   range-query-300-rand |     1.87752 |     1.67839 | -0.19912 |     ms |
|                                  90th percentile service time |   range-query-300-rand |     2.02667 |     1.99252 | -0.03415 |     ms |
|                                  99th percentile service time |   range-query-300-rand |     2.22245 |     2.72394 |  0.50149 |     ms |
| ... ||||||
|                                  50th percentile service time |  range-query-3000-rand |     1.83041 |     1.47596 | -0.35446 |     ms |
|                                  90th percentile service time |  range-query-3000-rand |     1.98185 |      1.8231 | -0.15875 |     ms |
|                                  99th percentile service time |  range-query-3000-rand |      2.1199 |     2.27615 |  0.15625 |     ms |
| ... ||||||
|                                  50th percentile service time | range-query-30000-rand |     1.79824 |     1.68221 | -0.11603 |     ms |
|                                  90th percentile service time | range-query-30000-rand |     1.92793 |     1.85027 | -0.07766 |     ms |
|                                  99th percentile service time | range-query-30000-rand |     2.07024 |     2.18596 |  0.11573 |     ms |

After running this, I think the savings in storage space by dropping the points data would outweigh the small potential performance gains, which might even be smaller in practice when caching etc... comes into play.

Commit:
Add helper methods to versions ScriptDocValues to help with version part extraction

Commit:
Merge branch 'master' into add-version-field

cbuescher commented:
@mayya-sharipova suggested that maybe we should add helpers to extract the meta-information that so far I store in subfields (if a version is a pre-release, major, minor, patch numbers) to the version fields ScriptDocValues implementation. I took a go at it in e4b0728 and while this will be slower than storing the information upfront, this is probably fast enough so use in the scenarios described on the issue that want to use this information (e.g. grouping on major version in aggregations of filtering non-prerelease versions only). I opted for "illegal" versions to not count as pre-releases and that all version numbers are returned as `null` for them, but this behaviour can be discussed of course. Adding the commit here for discussion, if we think this is helpful enough I'll also go ahead and drop the sub-fields.

Commit:
Removing points and subfields

cbuescher commented:
After discussions I ran more extended performance tests on a larger dataset of 10Mio docs and used fixed ranges that span aprroximately 5% (~500k docs), 10% (1Mio) nd 50% (5Mio) of the documents to get comparisons of longer running queries. I kept distinct runs on dictionary sizes of 300, 3000 and 30000 respectively. 
The first set of experiments compares range queries using only Postings (TermRangeQuery) as "baseline" and comparing to the current using Point/Docvalues (contender). The query edges are chosen so that versions fit into to 16 byte prefix we use for points to the query can take the shortcut of only using BinaryPoint.newRangeQuery and doesn't need an additional validation query. 
Looking at the 50th and 90th percentile, one can see "points" performing better the larger the dictionary and the query range gets. However, even for 30000 unique terms and 5 Mio queries documents, the speed-up is only in the order of 10%. For a medium sized dictionary (3000) the differences are less pronounced while it appears that for a small dictionary (300, thats a little more versions than ES currently has), the TermRangeQuery seems a bit faster on medium and large ranges.

|                                                        Metric |           Task |    Baseline |   Contender |     Diff |   Unit |
|--------------------------------------------------------------:|---------------:|------------:|------------:|---------:|-------:|
|                                  50th percentile service time |         300-5% |     17.6855 |     14.3997 | -3.28584 |     ms |
|                                  90th percentile service time |         300-5% |     18.5755 |     14.7387 | -3.83673 |     ms |
|                                                    ... |       |            |            |         |       |
|                                  50th percentile service time |        300-10% |     18.9116 |     22.7751 |  3.86351 |     ms |
|                                  90th percentile service time |        300-10% |     19.3731 |     23.3348 |  3.96168 |     ms |
|                                                    ... |       |            |            |         |       |
|                                  50th percentile service time |        300-50% |     68.6097 |     84.2955 |  15.6858 |     ms |
|                                  90th percentile service time |        300-50% |     76.4204 |     86.1675 |  9.74706 |     ms |
|                                                    ... |       |            |            |         |       |
|                                  50th percentile service time |        3000-5% |     11.8627 |     11.6883 | -0.17445 |     ms |
|                                  90th percentile service time |        3000-5% |     12.5302 |      11.952 | -0.57828 |     ms |
|                                                    ... |       |            |            |         |       |
|                                  50th percentile service time |       3000-10% |     19.3416 |     19.0441 | -0.29746 |     ms |
|                                  90th percentile service time |       3000-10% |     19.9481 |     19.9061 | -0.04193 |     ms |
|                                                    ... |       |            |            |         |       |
|                                  50th percentile service time |       3000-50% |     73.4131 |     75.3525 |  1.93946 |     ms |
|                                  90th percentile service time |       3000-50% |     75.1449 |     77.6721 |  2.52714 |     ms |
|                                                    ... |       |            |            |         |       |
|                                  50th percentile service time |       30000-5% |     13.8949 |     12.2752 | -1.61969 |     ms |
|                                  90th percentile service time |       30000-5% |     14.2412 |     12.5838 |  -1.6574 |     ms |
|                                                    ... |       |            |            |         |       |
|                                  50th percentile service time |      30000-10% |     22.3458 |     19.9012 | -2.44465 |     ms |
|                                  90th percentile service time |      30000-10% |     22.9008 |     20.3021 | -2.59865 |     ms |
|                                                    ... |       |            |            |         |       |
|                                  50th percentile service time |      30000-50% |     84.0917 |     74.6335 | -9.45822 |     ms |
|                                  90th percentile service time |      30000-50% |     86.3096 |     75.8234 | -10.4862 |     ms |

In the second experiment, the edges are chosen so an additional validation query is needed (e.g. by using an upper bound of "8.0.0-beta" while the number of returned documents are kept approximately the same as in the previous experiment. Here the query using points seems to always be slower, most likely due to the extra validation query on DocValues that is used here.

|                                                        Metric |           Task |    Baseline |   Contender |     Diff |   Unit |
|--------------------------------------------------------------:|---------------:|------------:|------------:|---------:|-------:|
|                                  50th percentile service time |    300-5%-edge |     17.1157 |     20.2967 |  3.18098 |     ms |
|                                  90th percentile service time |    300-5%-edge |     17.5806 |      20.624 |  3.04344 |     ms |
|                                                    ... |       |            |            |         |       
|                                  50th percentile service time |   300-10%-edge |     18.4925 |     33.6209 |  15.1284 |     ms |
|                                  90th percentile service time |   300-10%-edge |     18.9975 |      34.189 |  15.1915 |     ms |
|                                                    ... |       |            |            |         |       |
|                                  50th percentile service time |   300-50%-edge |     63.1801 |     122.593 |   59.413 |     ms |
|                                  90th percentile service time |   300-50%-edge |     64.7092 |     125.777 |  61.0679 |     ms |
|                                                    ... |       |            |            |         |       |
|                                  50th percentile service time |   3000-5%-edge |     11.8635 |     17.8834 |  6.01989 |     ms |
|                                  90th percentile service time |   3000-5%-edge |     12.2198 |     18.2252 |  6.00546 |     ms |
|                                                    ... |       |            |            |         |       |
|                                  50th percentile service time |  3000-10%-edge |       19.54 |     30.9377 |  11.3977 |     ms |
|                                  90th percentile service time |  3000-10%-edge |     20.1697 |     31.4531 |  11.2835 |     ms |
|                                                    ... |       |            |            |         |       |
|                                  50th percentile service time |  3000-50%-edge |     72.9496 |     113.946 |  40.9964 |     ms |
|                                  90th percentile service time |  3000-50%-edge |     75.5633 |     116.745 |  41.1814 |     ms |
|                                                    ... |       |            |            |         |       |
|                                  50th percentile service time |  30000-5%-edge |     13.7294 |      20.551 |   6.8216 |     ms |
|                                  90th percentile service time |  30000-5%-edge |     14.2017 |     20.9042 |  6.70249 |     ms |
|                                                    ... |       |            |            |         |       |
|                                  50th percentile service time | 30000-10%-edge |     22.5073 |      32.615 |  10.1077 |     ms |
|                                  90th percentile service time | 30000-10%-edge |     23.0239 |     39.8062 |  16.7823 |     ms |
|                                                    ... |       |            |            |         |       |
|                                  50th percentile service time | 30000-50%-edge |     83.8096 |     122.367 |  38.5572 |     ms |
|                                  90th percentile service time | 30000-50%-edge |     87.1256 |     125.348 |  38.2228 |     ms |

cbuescher commented:
An additional experiment comparing the performance of range queries on regular `keyword` fields (baseline) compared to the new version field using Postings shows little difference in performance as might be expected, since both use TermRangeQuery internally:

|                                                        Metric |           Task |    Baseline |   Contender |     Diff |   Unit |
|--------------------------------------------------------------:|---------------:|------------:|------------:|---------:|-------:|
|                                  50th percentile service time |         300-5% |     18.8628 |     17.4703 | -1.39255 |     ms |
|                                  90th percentile service time |         300-5% |     22.8014 |     18.2397 | -4.56168 |     ms |
|                                                    ... |       |            |            |         |       |
|                                  50th percentile service time |        300-10% |       16.13 |      16.505 |  0.37504 |     ms |
|                                  90th percentile service time |        300-10% |     17.6397 |     17.1114 | -0.52837 |     ms |
|                                                    ... |       |            |            |         |       |
|                                  50th percentile service time |        300-50% |     61.8667 |     57.0715 | -4.79513 |     ms |
|                                  90th percentile service time |        300-50% |     63.6463 |     60.0267 | -3.61953 |     ms |
|                                                    ... |       |            |            |         |       |
|                                  50th percentile service time |        3000-5% |     11.0996 |     11.0915 | -0.00811 |     ms |
|                                  90th percentile service time |        3000-5% |     11.3444 |     11.7128 |  0.36846 |     ms |
|                                                    ... |       |            |            |         |       |
|                                  50th percentile service time |       3000-10% |     16.2241 |     17.0877 |  0.86367 |     ms |
|                                  90th percentile service time |       3000-10% |     17.3865 |     17.5564 |  0.16988 |     ms |
|                                                    ... |       |            |            |         |       |
|                                  50th percentile service time |       3000-50% |     62.4286 |     60.8443 | -1.58435 |     ms |
|                                  90th percentile service time |       3000-50% |     64.8067 |      64.831 |  0.02429 |     ms |
|                                                    ... |       |            |            |         |       |
|                                  50th percentile service time |       30000-5% |     18.2399 |     18.0232 | -0.21662 |     ms |
|                                  90th percentile service time |       30000-5% |     18.5534 |     18.4378 | -0.11558 |     ms |
|                                                    ... |       |            |            |         |       |
|                                  50th percentile service time |      30000-10% |     29.2602 |     29.7478 |  0.48762 |     ms |
|                                  90th percentile service time |      30000-10% |     29.5937 |      30.745 |  1.15128 |     ms |
|                                                    ... |       |            |            |         |       |
|                                  50th percentile service time |      30000-50% |     108.254 |     114.964 |  6.71014 |     ms |
|                                  90th percentile service time |      30000-50% |     110.408 |     116.764 |  6.35637 |     ms |
|                                                    ... |       |            |            |         |       |

Commit:
Merge branch 'master' into add-version-field

cbuescher commented:
I went ahead and removed the usage of the points datastructure since this will reduce storage footprint while still allowing us to use the current exact matching and wildcard/regex implementation and the performance measurements above show that using points really speeds up the query only for large term dictionary sizes and querying millions of documents if the range edges fit into the 16 byte prefix. I'm happy to reconsider but wanted to move the PR into this direction since it reflects my current thinking.
Also removing indexing and testing for the subfields in favour of providing better access to this informatin via scripting.

Commit:
Add tests for new regex case-insensitivity option

Commit:
Adding docs

mayya-sharipova commented on docs/reference/mapping/types/keyword.asciidoc:
```diff
@@ -129,3 +129,5 @@ The following parameters are accepted by `keyword` fields:
 
 include::wildcard.asciidoc[]
+
+include::version.asciidoc[]
```
@cbuescher  I am curious why you considered `version` to be a part of keyword family? 

I would rather say that it should be a part of `structured` or `other` category?

Also curious what @jtibshirani  thinks as she worked on this organization.

cbuescher replied:
I put the docs here because we discussed the desired compatibility with the `keyword` field for the new field type a lot since users currently mostly seem to use `keyword` to indec these version constants. The `version` field should more or less behave like a keyword field, just with specialized semantics and ordering. I'm happy to move the docs anywhere else if there is another option that finds consensus.

mayya-sharipova replied:
thanks for the explanation, let's see what @jtibshirani thinks of this categorization

jtibshirani replied:
Thanks for looping me in, I was also imagining it would be listed under 'Structured'.

We recently added this concept of a 'field type family', which is a group of field types that have identical search behavior, but present different trade-offs (for example in terms of query speed). They are all present as the family type in field caps. The `version` field doesn't fit this description -- it has different search behavior on purpose, and doesn't present itself as a `keyword` in field caps.

cbuescher replied:
Thanks @jtibshirani I moved the docs in https://github.com/elastic/elasticsearch/pull/59773/commits/ccebd6f470c99c824a84211ed122ba10f1f9629c

mayya-sharipova commented on docs/reference/mapping/types/version.asciidoc:
```diff
@@ -0,0 +1,125 @@
+can e.g. be searched for exact matches using `match` or `term` queries and
+supports prefix and wildcard searches. The main benefit is that `range` queries
+will honour Semver ordering, so a `range` query between "1.0.0" and "1.5.0"
+will include versions of "1.2.3" but not "1.11.2" for example. Not that this
```
"Not that" => "Note that"?

mayya-sharipova commented on docs/reference/mapping/types/version.asciidoc:
```diff
@@ -0,0 +1,125 @@
+allowed (i.e. "1.2" or "1.2.3.4" qualify as valid versions while they wouldn't under
+strict Semver rules). Version strings that are not valid under the Semver definition
+(e.g. "1.2.alpha.4") can still be indexed and retrieved as exact matches, however they
+will all appear _after_ any valid version with regular alphabetical ordering. 
```
"with regular alphabetical ordering" =>  I am not quite sure what it means?  Are we sorting versions by an alphabetical ordering?

cbuescher replied:
Valid Semver versions like "1.2.3" are sorted numerically according to major, minor, patch, i.e. "1.2.1" < "1.11.1". There was a requirement to also be able to store and retrieve invalid versions like "1.a.123" for compatibility with existing solutions using "keyword", however we sort those after all valid versions and use the "default" keyword sorting which is alphabetical. Do the docs makes sense with this respect or do you think I can improve something?

mayya-sharipova replied:
@cbuescher Thanks for explanation, with it, the docs completely make sense,  I somehow assumed that "alphabetical ordering" applied to valid versions.

mayya-sharipova commented on docs/reference/mapping/types/version.asciidoc:
```diff
@@ -0,0 +1,125 @@
+
+[discrete]
+[[version-params]]
+==== Parameters for wildcard fields
```
wildcard fields => version fields?

mayya-sharipova commented on x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionFieldPlugin.java:
```diff
@@ -0,0 +1,39 @@
+import java.util.List;
+import java.util.Map;
+
+public class VersionFieldPlugin extends Plugin implements ActionPlugin, MapperPlugin {
```
Does this plugin need to implement `ActionPlugin`? It doesn't look like it provides new actions.

mayya-sharipova commented on x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java:
```diff
@@ -0,0 +1,432 @@
+
+    static class Builder extends ParametrizedFieldMapper.Builder {
+
+        private final Parameter<Map<String, String>> meta = Parameter.metaParam();
```
I am wondering  if we should also have  a `doc_values` parameter?
Since our `range` query doesn't depend on doc_values anymore, I am thinking may be a user can opt out of indexing `doc_values` if they don't need sorting, aggregation of this field, and script queries on pre-release, major, minor and patch

cbuescher replied:
I'm on the fence with this suggestion. I think the specialized sorting is one of the main features for using this field type, so I would assume most users wouldn't want to turn doc_values of. Maybe we can start without this option and add it if this is a really frequent ask (to save on storage spave while willing to pay the loss of functionality here)?

mayya-sharipova commented on docs/reference/mapping/types/version.asciidoc:
```diff
@@ -0,0 +1,125 @@
+the notable exception that more or less than three main version identifiers are
+allowed (i.e. "1.2" or "1.2.3.4" qualify as valid versions while they wouldn't under
+strict Semver rules). Version strings that are not valid under the Semver definition
+(e.g. "1.2.alpha.4") can still be indexed and retrieved as exact matches, however they
```
may be also add a note about empty strings, where they will be in sorted order?

cbuescher replied:
I can do that. Empty strings are considered as "invalid" version strings and are sorted after the valid ones, but before all other invalid versions.

mayya-sharipova commented on x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java:
```diff
@@ -0,0 +1,432 @@
+
+        @Override
+        public IndexFieldData.Builder fielddataBuilder(String fullyQualifiedIndexName, Supplier<SearchLookup> searchLookup) {
+            failIfNoDocValues();
```
if we decide to always index a field with doc values, this `failIfNoDocValues()` check seems unnecessary.

mayya-sharipova commented on x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java:
```diff
@@ -0,0 +1,432 @@
+
+        @Override
+        public Query rangeQuery(Object lowerTerm, Object upperTerm, boolean includeLower, boolean includeUpper, QueryShardContext context) {
+            if (context.allowExpensiveQueries() == false) {
```
from your last benchmarking tests, `rangeQuery` doesn't seem to be expensive, so this check to allow expensive queries looks unnecessary to me. WDYT?

cbuescher replied:
While I'm happy removing this check, the same check exists for keyword fields via StringFieldType#rangeQuery() which is also using TermRangeQuery under the hood. While we expect the dictionary size for version fields to be lower, we cannot be sure users are not (by accident) misusing this field to store hundreds of thousands of version strings, in which case performance should be similar to keyword field performance which we currently guard for by checking "allowExpensiveQueries()". Wdyt?

mayya-sharipova replied:
Thanks, I am also ok to have this check for consistency with `StringFieldType`.

jimczi replied:
Unlike keyword, range queries are the most important part of this field so I'd prefer that we ignore the `allowExpensiveQueries` in this field.

mayya-sharipova commented on x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java:
```diff
@@ -0,0 +1,432 @@
+                        + "' is set to false."
+                );
+            }
+            failIfNotIndexed();
```
Doesn't seem that a user has an option not to index a field. So this check looks unnecessary

mayya-sharipova commented on x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java:
```diff
@@ -0,0 +1,432 @@
+         * automaton in the query will assume unencoded terms. We are running through all terms, decode them and
+         * then run them through the automaton manually instead. This is not as efficient as intersecting the original
+         * Terms with the compiled automaton, but we expect the number of distinct version terms indexed into this field
+         * to be low enough and the use of "rexexp" queries on this field rare enough to brute-force this
```
nit: "rexexp"  -> "regexp"

mayya-sharipova commented on docs/reference/mapping/types/version.asciidoc:
```diff
@@ -0,0 +1,125 @@
+    Returns `true` if the field contains a pre-release version
+    
+getMajor()::
+    Returns the Major version for valid versions.
```
For `getMajor`, `getMinor`, and `getPatch` functions should we provide more details, e.g "returns Integer ... if a version is valid, or null otherwise..`

mayya-sharipova commented on docs/reference/mapping/types/version.asciidoc:
```diff
@@ -0,0 +1,125 @@
+    Returns `true` if the field contains a version thats legal according to the Semantic Versioning rules
+
+isPreRelease()::
+    Returns `true` if the field contains a pre-release version
```
should we also add that `false` is returned if a version is invalid? Or may be these functions should fail if the version is invalid than it will be up to a user to guard agains this.

cbuescher replied:
I'm thinking about switching the semantics here to `isRelease()` because that is more likely what users will filter for (e.g. show me all errors in released versions between 1.2.3 and 2.3.4). `isRelease()` then could be `false` for invalid versions and if you are interesed in the distinction you can always use `isValid()` with it. Would that make more sense to you?

cbuescher replied:
I changed this in f1bf52f, let me know if you think this is better or if I should revert.

mayya-sharipova replied:
Thanks, having `isRelease()` function with semantics you defined looks very good to me.

mayya-sharipova commented on x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionScriptDocValues.java:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    public String getValue() {
+        if (count == 0) {
```
For other script doc values, if a document doesn't have a value, we raise an error. I think we should follow the same standard here.

cbuescher replied:
Yes, this seems to be the behaviour for doc values on `keyword` field, so I'm going to switch to raising that error.

Commit:
Adressing review comments

mayya-sharipova commented:
@cbuescher Thanks for the benchmarking results, they look very good to me, and I agree with your conclusions of not indexing `points`.
I left more comments, and will continue the review tomorrow.

Commit:
Merge branch 'master' into add-version-field

Commit:
Switch script method from isPreRelease to isRelease

Commit:
Merge branch 'master' into add-version-field

mayya-sharipova commented on x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionEncoder.java:
```diff
@@ -0,0 +1,226 @@
+    private static final char PRERELEASE_SEPARATOR = '-';
+    private static final char DOT_SEPARATOR = '.';
+    private static final char BUILD_SEPARATOR = '+';
+    private static final String ENCODED_EMPTY_STRING = new String(new String(new byte[] { NO_PRERELEASE_SEPARATOR_BYTE }, Charsets.UTF_8));
```
`new String(new String(` -> should there be only a single `new String`?

cbuescher replied:
Heck yes!

Commit:
Another review iteration

mayya-sharipova commented on x-pack/plugin/versionfield/src/test/java/org/elasticsearch/xpack/versionfield/VersionStringFieldTests.java:
```diff
@@ -0,0 +1,498 @@
+        String indexName = setUpIndex("test");
+
+        // sort based on version field
+        SearchResponse response = client().prepareSearch(indexName)
```
Do you think we need to add some illegal versions and test that sort ASC will put them at the end?

cbuescher replied:
Sure can do.

mayya-sharipova commented on x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionFieldWildcardQuery.java:
```diff
@@ -0,0 +1,114 @@
+                    break;
+                case WILDCARD_CHAR:
+                    // this should also match leading digits, which have optional leading numeric marker and length bytes
+                    automata.add(optionalNumericCharPrefix());
```
@cbuescher This is a quite clever way to organize `wildcard` query!!! Good job here!!!

For this particular case of `?` case, I think we need to add additional optional automata of `PRERELEASE_SEPARATOR_BYTE` and `NO_PRERELEASE_SEPARATOR_BYTE`.  

Otherwise,  a query `"1.2.3??"` can't match  a version like `"1.2.3+b"` or a version like `"1.2.3-a"`

cbuescher replied:
Great catch, I'll add a test for this and will work on it

cbuescher commented:
@mayya-sharipova thanks, I hope I adressed your last comments in ce51484

mayya-sharipova commented on x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java:
```diff
@@ -0,0 +1,419 @@
+         * automaton in the query will assume unencoded terms. We are running through all terms, decode them and
+         * then run them through the automaton manually instead. This is not as efficient as intersecting the original
+         * Terms with the compiled automaton, but we expect the number of distinct version terms indexed into this field
+         * to be low enough and the use of "regexp" queries on this field rare enough to brute-force this
```
May be we can add to do for subsequent PRs to optimize these queries

cbuescher replied:
I'm not a bit fan of tracking backlog like this in TODOs in code, I'd rather open an issue to investigate improvements when time allows and track it with any other issue we have. Would that be okay for you?

mayya-sharipova replied:
@cbuescher Thanks, that would be fine.

cbuescher replied:
I opened https://github.com/elastic/elasticsearch/issues/62608 for tracking

mayya-sharipova reviewed (approved):
@cbuescher Thanks for all these iterations of addressing feedback and thanks for a great field type, this PR LGTM.

I have just left small comments

Commit:
Another round of reviews

Commit:
Moving docs under 'structured'

jimczi commented on docs/reference/mapping/types/version.asciidoc:
```diff
@@ -0,0 +1,129 @@
+--------------------------------------------------
+// TEST[continued]
+
+Functions available on via doc values in scripting are:
```
I understand how useful these functions can be but I'd prefer that we handle this in a follow up. It's unclear to me if these functions could be useful for other types as well (runtime or keyword field ?) so I think we should discuss this feature separately.

jimczi commented on x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java:
```diff
@@ -0,0 +1,526 @@
+
+        @Override
+        public Query prefixQuery(String value, MultiTermQuery.RewriteMethod method, QueryShardContext context) {
+            if (context.allowExpensiveQueries() == false) {
```
Can you remove the check please

jimczi commented on x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/ValidationOnSortedDv.java:
```diff
@@ -0,0 +1,130 @@
+ * Query that runs a validation for version ranges across sorted doc values.
+ * Used in conjunction with more selective query clauses.
+ */
+class ValidationOnSortedDv extends Query {
```
This class is not needed anymore ?

jimczi reviewed (approved):
The change looks good to me. I left one comment regarding the new script functions that I think should be discussed separately. I also wonder what family type this field should expose but that can also be decide in a follow up. So +1 to merge the current state since the main functionality we're after is there and consistent (accurate sort and range queries on Semver versions).

Commit:
Merge branch 'master' into add-version-field

Commit:
iter

Commit:
Removing specialized script doc values functions

cbuescher commented:
@mayya-sharipova @jimczi thanks for the reviews, I will work on backporting and open issues for follow ups where needed.
